src/coreclr/jit/lsraxarch.cpp

```diff
@@ -910,41 +910,142 @@ int LinearScan::BuildSelect(GenTreeOp* select)
+            BuildUse(cond);
+            srcCount++;
+        }
     }
 
-    // cmov family of instructions are special in that they only conditionally
-    // define the destination register, so when generating code for GT_SELECT
-    // we normally need to preface it by a move into the destination with one
-    // of the operands. We can avoid this if one of the operands is already in
-    // the destination register, so try to prefer that.
-    //
-    // Because of the above we also need to set delayRegFree on the intervals
-    // for contained operands. Otherwise we could pick a target register that
-    // conflicted with one of those registers.
-    //
-    if (select->gtOp1->isContained())
+    GenTree* trueVal  = select->gtOp1;
+    GenTree* falseVal = select->gtOp2;
+
+    RefPositionIterator op1UsesPrev = refPositions.backPosition();
+    assert(op1UsesPrev != refPositions.end());
+
+    RefPosition* uncontainedTrueRP = nullptr;
+    if (trueVal->isContained())
     {
-        srcCount += BuildDelayFreeUses(select->gtOp1);
+        srcCount += BuildOperandUses(trueVal);
     }
     else
     {
-        tgtPrefUse = BuildUse(select->gtOp1);
+        tgtPrefUse = uncontainedTrueRP = BuildUse(trueVal);
         srcCount++;
     }
 
-    if (select->gtOp2->isContained())
+    RefPositionIterator op2UsesPrev = refPositions.backPosition();
+
+    RefPosition* uncontainedFalseRP = nullptr;
+    if (falseVal->isContained())
     {
-        srcCount += BuildDelayFreeUses(select->gtOp2);
+        srcCount += BuildOperandUses(falseVal);
     }
     else
     {
-        tgtPrefUse2 = BuildUse(select->gtOp2);
+        tgtPrefUse2 = uncontainedFalseRP = BuildUse(falseVal);
         srcCount++;
     }
 
+    if ((tgtPrefUse != nullptr) && (tgtPrefUse2 != nullptr))
+    {
+        // CQ analysis shows that it's best to always prefer only the 'true'
+        // val here.
+        tgtPrefUse2 = nullptr;
+    }
+
+    // Codegen will emit something like:
+    //
+    // mov dstReg, falseVal
+    // cmov dstReg, trueVal
+    //
+    // We need to ensure that dstReg does not interfere with any register that
+    // appears in the second instruction. At the same time we want to
+    // preference the dstReg to be the same register as either falseVal/trueVal
+    // to be able to elide the mov whenever possible.
+    //
+    // While we could resolve the situation with either an internal register or
+    // by marking the uses as delay free unconditionally, this is a node used
+    // for very basic code patterns, so the logic here tries to be smarter to
+    // avoid the extra register pressure/potential copies.
+    //
+    // We have some flexibility as codegen can swap falseVal/trueVal as needed
+    // to avoid the conflict by reversing the sense of the cmov. If we can
+    // guarantee that the dstReg is used only in one of falseVal/trueVal, then
+    // we are good.
+    //
+    // To ensure the above we have some bespoke interference logic here on
```

Comment:
I wish there was an easy way to handle this particular scenario.

Reply:
Me too, but luckily this check didn't turn out to be too ugly.
I do wonder if we could use a similar trick in the interval building for RMW opers to avoid delay free in most cases, perhaps something to look at in a follow up.

Reply:
Well, we could but it would defeat the "linear" nature of the linear scan provided that it doesn't hurt much in common scenarios.

Reply:
Just to understand a bit more, `BuildDelayFreeUses()` doesn't achieve what you want? It doesn't delay free unconditionally, but does a check about matching interval, etc.?

Reply:
> it would defeat the "linear" nature of the linear scan

This does not affect the complexity since the number of ref positions created for op1/op2 is bounded by a constant (at most 2, for an indir with a contained LEA).

> Just to understand a bit more, BuildDelayFreeUses() doesn't achieve what you want? It doesn't delay free unconditionally, but does a check about matching interval, etc.?

I did start out with `BuildDelayFreeUses`, but the extra flexibility we have in codegen (by swapping the trueVal/falseVal and reversing the condition) means that we don't need to mark something delay-free in practically all situations. `BuildDelayFreeUses` handles only situations where the other operand is a local, but for this particular node it could also be another contained indirection.

I looked at the RMW logic and I don't think it benefits from doing this since for the RMW opers we don't have the capability to contain two indirections anyway.